packages/hardhat-core/src/internal/core/config/config-loading.ts

```diff
@@ -109,7 +109,7 @@ export function loadConfigAndTasks(
     checkUnsupportedRemappings(resolved);
   }
 
-  return resolved;
+  return { ...resolved, userConfig: frozenUserConfig };
```

Comment:
This seems redundant, since `userConfig` is already added by `resolveConfig`. That is: if I undo this line, then it still compiles fine.

If we _want_ `config.userConfig` to be frozen, then I would pass it as a third parameter to `resolveConfig` and use it there.